redis/src/cluster_handling/async_connection/mod.rs

```diff
@@ -172,30 +170,62 @@ where
         #[cfg(feature = "cache-aio")]
         let cache_manager = cluster_params.cache_manager.clone();
         let runtime = Runtime::locate();
-        ClusterConnInner::new(initial_nodes, cluster_params)
-            .await
-            .map(|inner| {
-                let (sender, mut receiver) = mpsc::channel::<Message<_>>(100);
-                let stream = async move {
-                    let _ = stream::poll_fn(move |cx| receiver.poll_recv(cx))
-                        .map(Ok)
-                        .forward(inner)
-                        .await;
-                };
-                let _task_handle = HandleContainer::new(runtime.spawn(stream));
+        let inner = ClusterConnInner::new(initial_nodes, cluster_params);
+        let inner = inner.wait_for_initial_connection().await?;
+
+        let (sender, mut receiver) = mpsc::channel::<Message<_>>(100);
+        let stream = async move {
+            let _ = stream::poll_fn(move |cx| receiver.poll_recv(cx))
+                .map(Ok)
+                .forward(inner)
+                .await;
+        };
+        let _task_handle = HandleContainer::new(runtime.spawn(stream));
+
+        Ok(ClusterConnection {
+            sender,
+            state: Arc::new(ClientSideState {
+                protocol,
+                _task_handle,
+                response_timeout,
+                runtime,
+                #[cfg(feature = "cache-aio")]
+                cache_manager,
+            }),
+        })
+    }
 
-                ClusterConnection {
-                    sender,
-                    state: Arc::new(ClientSideState {
-                        protocol,
-                        _task_handle,
-                        response_timeout,
-                        runtime,
-                        #[cfg(feature = "cache-aio")]
-                        cache_manager,
-                    }),
-                }
-            })
+    pub(crate) fn new_pending(
+        initial_nodes: &[ConnectionInfo],
+        cluster_params: ClusterParams,
+    ) -> ClusterConnection<C> {
+        let protocol = cluster_params.protocol.unwrap_or_default();
```

Comment:
since this is almost completely a repetition of `new`, I'd prefer it if the logic was consolidated into a single function, with arguments that describe the difference in behavior.

Reply:
The two function differ in that one is async and the other (used for `get_pending_`) isn't , I don't see a way to combine them into one (without forcing `get_pending_` to be async for no reason).